### ISSUE TITLE
Add support for messageId

### ISSUE THREAD

**Issue by WaleedAshraf** [asyncapi/spec#458]
**Is your feature request related to a problem? Please describe.**
Currently, there is no support for a unique identifier for messages. My use case is to use `messageId` for validation.

**Describe the solution you'd like**
We have [operationId](https://asyncapi.io/docs/specifications/2.0.0#operationTraitObject). `messageId` would be the exact same field for messages.

**Describe alternatives you've considered**
For now, I'm relying on the `name` field and adding extra checks on it. But it's not required to be unique by specs.

**Additional context**
Related slack conv: https://asyncapi.slack.com/archives/C34F2JV0U/p1604399108296600
Another previous conv about adding messageId: https://github.com/asyncapi/asyncapi/issues/94#issuecomment-576433439
Related request from tooling: https://github.com/WaleedAshraf/asyncapi-validator/issues/25

**Comment by derberg** [comment]
@WaleedAshraf better copy-paste the slack conversation as we are on slack free plan and it will soon be lost

**Comment by WaleedAshraf** [comment]
Yup, I also noticed that, when trying to search for an old conv.

![image](https://user-images.githubusercontent.com/8335457/98115105-92ba0100-1ea6-11eb-910d-46910fae6469.png)

**Comment by derberg** [comment]
@fmvilas I think we can add label that indicates this issue reached `Stage 1`. What do you think. @WaleedAshraf just has to open a new PR as the other one was closed some time ago, but was already approved.

@WaleedAshraf keep in mind we no longer have schema in spec repo but now it is only in https://github.com/asyncapi/asyncapi-node

to reach `Stage 2` we are missing a PR on the JS Parser that enables the feature. Not everything can be done with JSON Schema. To assure uniqueness of IDs, in my opinion there needs to be a custom validation provided as we did with operationId.

**Comment by fmvilas** [comment]
Yeah, agree this one is on Stage 1. I've added the proper label.

**Comment by WaleedAshraf** [comment]
@derberg in asyncapi-node I already have a pr https://github.com/asyncapi/asyncapi-node/pull/24
Do we still need a PR here? I'm still confused about how schema changes take place.
Maybe we can document the whole process of what needs to be done in which repo against which branch, etc.

**Comment by WaleedAshraf** [comment]
I think I should create PR against 2021-06-release branch. 🤔

**Comment by derberg** [comment]
@WaleedAshraf 

> I think we need PR against the "2021-06-release" branch.

every PR for 2.1 release must be against `2021-06-release`

> Also, do we need to update js-parser right away? It doesn't need to be? Otherwise, we can end up with the same problem that other libs/tooling blocking newer versions of the schema.

To get a change in the spec it must reach Stage 3 and it requires a change (approved PR) against the release branch. Have a look at the [guide](https://github.com/asyncapi/spec/blob/master/CONTRIBUTING.md#stage-3-accepted)

We want to have at least schema and parser supporting the latest spec as the most important tools. The others can follow up independently before or after the release.

**Comment by WaleedAshraf** [comment]
Got it. I don't think I'll have time any soon. So, I'll just leave it here for someone to pick and I can help out if needed.

**Comment by joerg-walter-de** [comment]
I am coming from me looking for a message-specific operationId. 

I am used to generate client code for OpenAPI specs, that is for example callable JS functions where the function names match the operationId s and the functions provide datagram / parameter validation. And obviously it would be nice for a client to be able to call sendPing(), subscribeToData() etc. that is, to have a send-function for each possible message. Having a operationId for each message would help achieving that, I guess.

Currently operationId is bound to a channel. But since it may very well be reasonable to  send multiple types of messages through the same channel and since it would be good to have a message specific send-function and a message-specific handler function for incoming messages a message-specific ID but probably even a message-specific operation ID may make sense.

With OpenAPI each path/method combination must have its own operation. There is an inbalance in the hierarchy in AsyncAPI: There is no equivalent for an operation let's say for create entry.

**Comment by smoya** [comment]
I'm moving back this to `RFC 0` as there is no PR with spec edits. Also added `Needs Champion` label.

**Comment by smoya** [comment]
Hi @WaleedAshraf! since this RFC has been here for some time, I'm wondering if you would like to retake it and be the champion, hopefully, you have some time now.

**Comment by WaleedAshraf** [comment]
@smoya As I remember from the last discussion, this will require changes at multiple places and now I don't have the full context. I think you can help with defining each task as a separate issue, which can be picked up by anyone. I might be able to pick those.

**Comment by derberg** [comment]
change needs to be done in repo where JSON Schema is and proper helper function added to JS Parser. Should not be much work to do IMHO, we can help coordinate for sure.

**Comment by smoya** [comment]
@WaleedAshraf As you requested, here is the list of issues so you can track progress:

- https://github.com/asyncapi/spec-json-schemas/issues/131
- https://github.com/asyncapi/parser-js/issues/414

**Comment by dalelane** [comment]
@WaleedAshraf Would you like this one included in the [2.3 release](https://github.com/asyncapi/spec/issues/675)? I know it's been quiet for a little while, but it looks like it could be do-able - I'd be happy to help if you'd like.

**Comment by WaleedAshraf** [comment]
Hi @dalelane
I don't think I'll have time anytime soon. :(

**Comment by smoya** [comment]
@WaleedAshraf We plan to release [2.4 release](https://github.com/asyncapi/spec/issues/735) soon (April 2022). Do you think you will have some time for materializing this RFC?
I can help with it.

**Comment by WaleedAshraf** [comment]
@smoya Yes, I'll have some time. Let's sync on this on Slack.

**Comment by fmvilas** [comment]
Is Waleed back? 😱 🚀 🎉

**Comment by WaleedAshraf** [comment]
@smoya PR is there 🎉 . Who can review it?

**Comment by derberg** [comment]
@WaleedAshraf we also need an update to JSON Schema, you had a PR in the past https://github.com/asyncapi/spec-json-schemas/pull/24 but you just need to create new against proper file and branch

**Comment by WaleedAshraf** [comment]
@derberg Yes. I will do that once the PR for parser PR is approved. As there can be more changes required and it gets delayed, etc.

**Comment by WaleedAshraf** [comment]
@derberg Here https://github.com/asyncapi/spec-json-schemas/pull/195

**Comment by derberg** [comment]
I think we can move this one to `Stage 2: Draft`
- we have https://github.com/asyncapi/spec-json-schemas/pull/195
- and https://github.com/asyncapi/parser-js/pull/510

and also solution itself makes a lot of sense and can't really be done differently.
@fmvilas @dalelane 

cc @smoya

**Comment by fmvilas** [comment]
Yeah, let's move it to Stage 2 👍
